File: artifacts/api-server/data_loader.py
"""
Data loader — reads all PSL Excel files from attached_assets/.
Uses pandas + openpyxl; no AI logic here.
"""
import os
import pandas as pd

DATA_DIR = os.path.join(os.path.dirname(__file__), "..", "..", "attached_assets")

_cached: dict | None = None


def _p(fname: str) -> str:
    return os.path.join(DATA_DIR, fname)


def _to_num(v) -> float:
    if v is None or (isinstance(v, float) and str(v) == "nan"):
        return 0.0
    try:
        return float(str(v).strip().replace("-", "0") or "0")
    except Exception:
        return 0.0


def _to_str(v) -> str:
    if v is None or (isinstance(v, float) and str(v) == "nan"):
        return ""
    return str(v).strip()


def load_dataset() -> dict:
    global _cached
    if _cached is not None:
        return _cached

    # ── Master Dataset ────────────────────────────────────────────────────────
    master_df = pd.read_excel(
        _p("PSL_Master_Dataset_1777931713613.xlsx"),
        sheet_name="Master Dataset",
        header=1,   # row 0 is title, row 1 is headers
        dtype=str,
    )
    master_records = []
    for _, row in master_df.iterrows():
        player = _to_str(row.get("player", ""))
        if not player:
            continue
        master_records.append({
            "player":            player,
            "role":              _to_str(row.get("role", "")),
            "bowlingType":       _to_str(row.get("bowling_type", "")),
            "date":              _to_str(row.get("date", "")),
            "venue":             _to_str(row.get("venue", "")),
            "opposition":        _to_str(row.get("opposition", "")),
            "runs":              _to_num(row.get("runs", 0)),
            "balls":             _to_num(row.get("balls", 0)),
            "fours":             _to_num(row.get("fours", 0)),
            "sixes":             _to_num(row.get("sixes", 0)),
            "overs":             _to_num(row.get("overs", 0)),
            "ballsBowled":       _to_num(row.get("balls_bowled", 0)),
            "runsConceded":      _to_num(row.get("runs_conceded", 0)),
            "wickets":           _to_num(row.get("wickets", 0)),
            "battingStrikeRate": _to_num(row.get("batting_strike_rate", 0)),
            "economy":           _to_num(row.get("economy", 0)),
            "battingPosition":   _to_num(row.get("batting_position", 0)),
        })

    # ── Player Roles Master ───────────────────────────────────────────────────
    roles_df = pd.read_excel(
        _p("PSL_Pakistani_Players_Roles_Master_1777931713617.xlsx"),
        sheet_name=0,
        dtype=str,
    )
    player_roles = []
    for _, row in roles_df.iterrows():
        player = _to_str(row.get("Player", ""))
        if not player:
            continue
        player_roles.append({
            "player":        player,
            "role":          _to_str(row.get("Role", "")),
            "span":          _to_str(row.get("Span", "")),
            "batMatches":    _to_num(row.get("Bat Matches", 0)),
            "batInnings":    _to_num(row.get("Bat Innings", 0)),
            "notOut":        _to_num(row.get("Not Out", 0)),
            "runs":          _to_num(row.get("Runs", 0)),
            "highestScore":  _to_str(row.get("Highest Score", "0")),
            "batAverage":    _to_num(row.get("Bat Average", 0)),
            "ballsFaced":    _to_num(row.get("Balls Faced", 0)),
            "batStrikeRate": _to_num(row.get("Bat Strike Rate", 0)),
            "hundreds":      _to_num(row.get("100s", 0)),
            "fifties":       _to_num(row.get("50s", 0)),
            "ducks":         _to_num(row.get("Ducks", 0)),
            "fours":         _to_num(row.get("4s", 0)),
            "sixes":         _to_num(row.get("6s", 0)),
            "bowlMatches":   _to_num(row.get("Bowl Matches", 0)),
            "bowlInnings":   _to_num(row.get("Bowl Innings", 0)),
            "overs":         _to_num(row.get("Overs", 0)),
            "maidens":       _to_num(row.get("Maidens", 0)),
            "runsConceded":  _to_num(row.get("Runs Conceded", 0)),
            "wickets":       _to_num(row.get("Wickets", 0)),
            "bowlAverage":   _to_num(row.get("Bowl Average", 0)),
            "economyRate":   _to_num(row.get("Economy Rate", 0)),
            "bowlStrikeRate":_to_num(row.get("Bowl Strike Rate", 0)),
        })

    # ── Ground Bowling Stats ──────────────────────────────────────────────────
    gbowl_df = pd.read_excel(
        _p("PSL_Pakistani_Ground_Bowling_1777931713615.xlsx"),
        sheet_name=0,
        dtype=str,
    )
    ground_bowling = []
    for _, row in gbowl_df.iterrows():
        ground = _to_str(row.get("Ground/Venue", ""))
        player = _to_str(row.get("Player", ""))
        if not ground or not player:
            continue
        ground_bowling.append({
            "ground":     ground,
            "player":     player,
            "span":       _to_str(row.get("Span", "")),
            "matches":    _to_num(row.get("Matches", 0)),
            "innings":    _to_num(row.get("Innings", 0)),
            "overs":      _to_num(row.get("Overs", 0)),
            "wickets":    _to_num(row.get("Wickets", 0)),
            "average":    _to_num(row.get("Average", 0)),
            "economy":    _to_num(row.get("Economy", 0)),
            "strikeRate": _to_num(row.get("Strike Rate", 0)),
        })

    # ── Ground Batting Stats (4 sheets) ──────────────────────────────────────
    ground_batting: dict[str, list] = {}
    xl = pd.ExcelFile(_p("psl_batters_ground_stats_1777931713609.xlsx"))
    for sheet in xl.sheet_names:
        df = pd.read_excel(xl, sheet_name=sheet, header=1, dtype=str)
        records = []
        for _, row in df.iterrows():
            player = _to_str(row.get("Player", ""))
            if not player:
                continue
            records.append({
                "venue":        sheet,
                "player":       player,
                "matches":      _to_num(row.get("Mat", 0)),
                "innings":      _to_num(row.get("Inns", 0)),
                "notOut":       _to_num(row.get("NO", 0)),
                "runs":         _to_num(row.get("Runs", 0)),
                "highestScore": _to_str(row.get("HS", "0")),
                "average":      _to_num(row.get("Ave", 0)),
                "ballsFaced":   _to_num(row.get("BF", 0)),
                "strikeRate":   _to_num(row.get("SR", 0)),
                "hundreds":     _to_num(row.get("100", 0)),
                "fifties":      _to_num(row.get("50", 0)),
                "ducks":        _to_num(row.get("0", 0)),
            })
        ground_batting[sheet] = records

    # ── All-Rounders ──────────────────────────────────────────────────────────
    wk_xl = pd.ExcelFile(_p("psl_wicketkeepers_fielders_and_allrounder_1777931713618.xlsx"))
    ar_df = pd.read_excel(wk_xl, sheet_name="All-rounders", header=1, dtype=str)
    all_rounders = []
    for _, row in ar_df.iterrows():
        player = _to_str(row.get("Player", ""))
        if not player:
            continue
        all_rounders.append({
            "player":       player,
            "span":         _to_str(row.get("Span", "")),
            "matches":      _to_num(row.get("Mat", 0)),
            "runs":         _to_num(row.get("Runs", 0)),
            "highestScore": _to_str(row.get("HS", "0")),
            "batAvg":       _to_num(row.get("Bat Avg", 0)),
            "batSR":        _to_num(row.get("Bat SR", 0)),
            "wickets":      _to_num(row.get("Wkts", 0)),
            "best":         _to_str(row.get("Best", "")),
            "bowlAvg":      _to_num(row.get("Bowl Avg", 0)),
            "bowlEco":      _to_num(row.get("Bowl Eco", 0)),
        })

    # ── Wicketkeepers ─────────────────────────────────────────────────────────
    wk_df = pd.read_excel(wk_xl, sheet_name="WK Batting", header=1, dtype=str)
    wicketkeepers = []
    for _, row in wk_df.iterrows():
        player = _to_str(row.get("Player", ""))
        if not player:
            continue
        wicketkeepers.append({
            "player":       player,
            "span":         _to_str(row.get("Span", "")),
            "matches":      _to_num(row.get("Mat", 0)),
            "innings":      _to_num(row.get("Inns", 0)),
            "notOut":       _to_num(row.get("NO", 0)),
            "runs":         _to_num(row.get("Runs", 0)),
            "highestScore": _to_str(row.get("HS", "0")),
            "average":      _to_num(row.get("Ave", 0)),
            "ballsFaced":   _to_num(row.get("BF", 0)),
            "strikeRate":   _to_num(row.get("SR", 0)),
        })

    # ── Combined Bowlers ──────────────────────────────────────────────────────
    bowlers_df = pd.read_excel(
        _p("PSL_Pakistani_Bowlers_Combined_1777931713614.xlsx"),
        sheet_name=0,
        dtype=str,
    )
    bowlers = []
    for _, row in bowlers_df.iterrows():
        player = _to_str(row.get("Player", ""))
        if not player:
            continue
        bowlers.append({
            "player":      player,
            "bowlerType":  _to_str(row.get("Bowler Type", "")),
            "span":        _to_str(row.get("Span", "")),
            "matches":     _to_num(row.get("Matches", 0)),
            "innings":     _to_num(row.get("Innings", 0)),
            "overs":       _to_num(row.get("Overs", 0)),
            "wickets":     _to_num(row.get("Wickets", 0)),
            "average":     _to_num(row.get("Average", 0)),
            "economy":     _to_num(row.get("Economy", 0)),
            "strikeRate":  _to_num(row.get("Strike Rate", 0)),
        })

    _cached = {
        "masterRecords": master_records,
        "playerRoles":   player_roles,
        "groundBowling": ground_bowling,
        "groundBatting": ground_batting,
        "allRounders":   all_rounders,
        "wicketkeepers": wicketkeepers,
        "bowlers":       bowlers,
    }
    return _cached

File: artifacts/api-server/src/lib/dataLoader.ts
import XLSX from "xlsx";
import path from "path";

// Use process.cwd() which resolves to artifacts/api-server when the server runs,
// so ../../attached_assets correctly points to workspace/attached_assets
const DATA_DIR = path.join(process.cwd(), "../../attached_assets");

export interface MasterRecord {
  player: string;
  role: string;
  bowlingType: string;
  date: string;
  venue: string;
  opposition: string;
  runs: number;
  balls: number;
  fours: number;
  sixes: number;
  overs: number;
  ballsBowled: number;
  runsConceded: number;
  wickets: number;
  battingStrikeRate: number;
  economy: number;
  battingPosition: number;
}

export interface PlayerRoleMaster {
  player: string;
  role: string;
  span: string;
  batMatches: number;
  batInnings: number;
  notOut: number;
  runs: number;
  highestScore: string;
  batAverage: number;
  ballsFaced: number;
  batStrikeRate: number;
  hundreds: number;
  fifties: number;
  ducks: number;
  fours: number;
  sixes: number;
  bowlMatches: number;
  bowlInnings: number;
  overs: number;
  maidens: number;
  runsConceded: number;
  wickets: number;
  bowlAverage: number;
  economyRate: number;
  bowlStrikeRate: number;
}

export interface GroundBowlingRecord {
  ground: string;
  player: string;
  span: string;
  matches: number;
  innings: number;
  overs: number;
  wickets: number;
  average: number;
  economy: number;
  strikeRate: number;
}

export interface GroundBattingRecord {
  venue: string;
  player: string;
  matches: number;
  innings: number;
  notOut: number;
  runs: number;
  highestScore: string;
  average: number;
  ballsFaced: number;
  strikeRate: number;
  hundreds: number;
  fifties: number;
  ducks: number;
}

export interface AllRounderRecord {
  player: string;
  span: string;
  matches: number;
  runs: number;
  highestScore: string;
  batAvg: number;
  batSR: number;
  wickets: number;
  best: string;
  bowlAvg: number;
  bowlEco: number;
}

export interface WKRecord {
  player: string;
  span: string;
  matches: number;
  innings: number;
  notOut: number;
  runs: number;
  highestScore: string;
  average: number;
  ballsFaced: number;
  strikeRate: number;
}

export interface CombinedBowlerRecord {
  player: string;
  bowlerType: string;
  span: string;
  matches: number;
  innings: number;
  overs: number;
  wickets: number;
  average: number;
  economy: number;
  strikeRate: number;
}

export interface PSLDataset {
  masterRecords: MasterRecord[];
  playerRoles: PlayerRoleMaster[];
  groundBowling: GroundBowlingRecord[];
  groundBatting: Map<string, GroundBattingRecord[]>; // venue -> records
  allRounders: AllRounderRecord[];
  wicketkeepers: WKRecord[];
  bowlers: CombinedBowlerRecord[];
}

function readFile(filename: string): XLSX.WorkBook {
  return XLSX.readFile(path.join(DATA_DIR, filename));
}

function toNum(v: unknown): number {
  if (v === null || v === undefined || v === "" || v === "-") return 0;
  const n = Number(v);
  return isNaN(n) ? 0 : n;
}

function toStr(v: unknown): string {
  if (v === null || v === undefined) return "";
  return String(v).trim();
}

let cachedDataset: PSLDataset | null = null;

export function loadDataset(): PSLDataset {
  if (cachedDataset) return cachedDataset;

  // --- Master Dataset ---
  const masterWb = readFile("PSL_Master_Dataset_1777931713613.xlsx");
  const masterSheet = masterWb.Sheets["Master Dataset"];
  const masterRaw = XLSX.utils.sheet_to_json<Record<string, unknown>>(masterSheet, { header: 1 });
  // Row 0 is title, row 1 is headers, rows 2+ are data
  const masterHeaders = masterRaw[1] as string[];
  const masterRecords: MasterRecord[] = [];
  for (let i = 2; i < masterRaw.length; i++) {
    const row = masterRaw[i] as unknown[];
    if (!row || !row[0]) continue;
    const getCol = (name: string) => row[masterHeaders.indexOf(name)];
    masterRecords.push({
      player: toStr(getCol("player")),
      role: toStr(getCol("role")),
      bowlingType: toStr(getCol("bowling_type")),
      date: toStr(getCol("date")),
      venue: toStr(getCol("venue")),
      opposition: toStr(getCol("opposition")),
      runs: toNum(getCol("runs")),
      balls: toNum(getCol("balls")),
      fours: toNum(getCol("fours")),
      sixes: toNum(getCol("sixes")),
      overs: toNum(getCol("overs")),
      ballsBowled: toNum(getCol("balls_bowled")),
      runsConceded: toNum(getCol("runs_conceded")),
      wickets: toNum(getCol("wickets")),
      battingStrikeRate: toNum(getCol("batting_strike_rate")),
      economy: toNum(getCol("economy")),
      battingPosition: toNum(getCol("batting_position")),
    });
  }

  // --- Player Roles Master ---
  const rolesWb = readFile("PSL_Pakistani_Players_Roles_Master_1777931713617.xlsx");
  const rolesSheet = rolesWb.Sheets[rolesWb.SheetNames[0]];
  const rolesRaw = XLSX.utils.sheet_to_json<Record<string, unknown>>(rolesSheet);
  const playerRoles: PlayerRoleMaster[] = rolesRaw
    .filter((r) => r["Player"])
    .map((r) => ({
      player: toStr(r["Player"]),
      role: toStr(r["Role"]),
      span: toStr(r["Span"]),
      batMatches: toNum(r["Bat Matches"]),
      batInnings: toNum(r["Bat Innings"]),
      notOut: toNum(r["Not Out"]),
      runs: toNum(r["Runs"]),
      highestScore: toStr(r["Highest Score"]),
      batAverage: toNum(r["Bat Average"]),
      ballsFaced: toNum(r["Balls Faced"]),
      batStrikeRate: toNum(r["Bat Strike Rate"]),
      hundreds: toNum(r["100s"]),
      fifties: toNum(r["50s"]),
      ducks: toNum(r["Ducks"]),
      fours: toNum(r["4s"]),
      sixes: toNum(r["6s"]),
      bowlMatches: toNum(r["Bowl Matches"]),
      bowlInnings: toNum(r["Bowl Innings"]),
      overs: toNum(r["Overs"]),
      maidens: toNum(r["Maidens"]),
      runsConceded: toNum(r["Runs Conceded"]),
      wickets: toNum(r["Wickets"]),
      bowlAverage: toNum(r["Bowl Average"]),
      economyRate: toNum(r["Economy Rate"]),
      bowlStrikeRate: toNum(r["Bowl Strike Rate"]),
    }));

  // --- Ground Bowling Stats ---
  const groundBowlWb = readFile("PSL_Pakistani_Ground_Bowling_1777931713615.xlsx");
  const groundBowlSheet = groundBowlWb.Sheets[groundBowlWb.SheetNames[0]];
  const groundBowlRaw = XLSX.utils.sheet_to_json<Record<string, unknown>>(groundBowlSheet);
  const groundBowling: GroundBowlingRecord[] = groundBowlRaw
    .filter((r) => r["Ground/Venue"] && r["Player"])
    .map((r) => ({
      ground: toStr(r["Ground/Venue"]),
      player: toStr(r["Player"]),
      span: toStr(r["Span"]),
      matches: toNum(r["Matches"]),
      innings: toNum(r["Innings"]),
      overs: toNum(r["Overs"]),
      wickets: toNum(r["Wickets"]),
      average: toNum(r["Average"]),
      economy: toNum(r["Economy"]),
      strikeRate: toNum(r["Strike Rate"]),
    }));

  // --- Ground Batting Stats (multiple sheets: Karachi, Multan, Lahore, Rawalpindi) ---
  const groundBattingWb = readFile("psl_batters_ground_stats_1777931713609.xlsx");
  const groundBatting = new Map<string, GroundBattingRecord[]>();
  for (const sheetName of groundBattingWb.SheetNames) {
    const sheet = groundBattingWb.Sheets[sheetName];
    const raw = XLSX.utils.sheet_to_json<unknown[]>(sheet, { header: 1 }) as unknown[][];
    // Row 0 is title, row 1 is headers, rows 2+ are data
    const headers = raw[1] as string[];
    if (!headers) continue;
    const records: GroundBattingRecord[] = [];
    for (let i = 2; i < raw.length; i++) {
      const row = raw[i] as unknown[];
      if (!row || !row[0]) continue;
      const g = (name: string) => row[headers.indexOf(name)];
      records.push({
        venue: sheetName,
        player: toStr(g("Player")),
        matches: toNum(g("Mat")),
        innings: toNum(g("Inns")),
        notOut: toNum(g("NO")),
        runs: toNum(g("Runs")),
        highestScore: toStr(g("HS")),
        average: toNum(g("Ave")),
        ballsFaced: toNum(g("BF")),
        strikeRate: toNum(g("SR")),
        hundreds: toNum(g("100")),
        fifties: toNum(g("50")),
        ducks: toNum(g("0")),
      });
    }
    groundBatting.set(sheetName, records);
  }

  // --- All-Rounders ---
  const wkWb = readFile("psl_wicketkeepers_fielders_and_allrounder_1777931713618.xlsx");
  const arSheet = wkWb.Sheets["All-rounders"];
  const arRaw = XLSX.utils.sheet_to_json<unknown[]>(arSheet, { header: 1 }) as unknown[][];
  const arHeaders = arRaw[1] as string[];
  const allRounders: AllRounderRecord[] = [];
  for (let i = 2; i < arRaw.length; i++) {
    const row = arRaw[i] as unknown[];
    if (!row || !row[0]) continue;
    const g = (name: string) => row[arHeaders.indexOf(name)];
    allRounders.push({
      player: toStr(g("Player")),
      span: toStr(g("Span")),
      matches: toNum(g("Mat")),
      runs: toNum(g("Runs")),
      highestScore: toStr(g("HS")),
      batAvg: toNum(g("Bat Avg")),
      batSR: toNum(g("Bat SR")),
      wickets: toNum(g("Wkts")),
      best: toStr(g("Best")),
      bowlAvg: toNum(g("Bowl Avg")),
      bowlEco: toNum(g("Bowl Eco")),
    });
  }

  // --- Wicketkeepers ---
  const wkSheet = wkWb.Sheets["WK Batting"];
  const wkRaw = XLSX.utils.sheet_to_json<unknown[]>(wkSheet, { header: 1 }) as unknown[][];
  const wkHeaders = wkRaw[1] as string[];
  const wicketkeepers: WKRecord[] = [];
  for (let i = 2; i < wkRaw.length; i++) {
    const row = wkRaw[i] as unknown[];
    if (!row || !row[0]) continue;
    const g = (name: string) => row[wkHeaders.indexOf(name)];
    wicketkeepers.push({
      player: toStr(g("Player")),
      span: toStr(g("Span")),
      matches: toNum(g("Mat")),
      innings: toNum(g("Inns")),
      notOut: toNum(g("NO")),
      runs: toNum(g("Runs")),
      highestScore: toStr(g("HS")),
      average: toNum(g("Ave")),
      ballsFaced: toNum(g("BF")),
      strikeRate: toNum(g("SR")),
    });
  }

  // --- Combined Bowlers ---
  const bowlersWb = readFile("PSL_Pakistani_Bowlers_Combined_1777931713614.xlsx");
  const bowlersSheet = bowlersWb.Sheets[bowlersWb.SheetNames[0]];
  const bowlersRaw = XLSX.utils.sheet_to_json<Record<string, unknown>>(bowlersSheet);
  const bowlers: CombinedBowlerRecord[] = bowlersRaw
    .filter((r) => r["Player"])
    .map((r) => ({
      player: toStr(r["Player"]),
      bowlerType: toStr(r["Bowler Type"]),
      span: toStr(r["Span"]),
      matches: toNum(r["Matches"]),
      innings: toNum(r["Innings"]),
      overs: toNum(r["Overs"]),
      wickets: toNum(r["Wickets"]),
      average: toNum(r["Average"]),
      economy: toNum(r["Economy"]),
      strikeRate: toNum(r["Strike Rate"]),
    }));

  cachedDataset = {
    masterRecords,
    playerRoles,
    groundBowling,
    groundBatting,
    allRounders,
    wicketkeepers,
    bowlers,
  };

  return cachedDataset;
}
